src/modules/flow/grove/grove.c

```diff
@@ -660,101 +675,150 @@ write_string(struct lcd_data *mdata, char *str)
 #define LCD_COLOR_B (2)
 
 #define TIME_TO_CLEAR (2)
+#define I2C_STEP_TIME (1)
+#define TIME_TO_TURN_ON (55)
 
-#define I2C_STEP_TIME 1
+static int command_queue_process(struct lcd_data *mdata);
 
-static bool command_queue_process(void *data);
+static bool
+timer_cb(void *data)
+{
+    int r;
+    struct lcd_data *mdata = data;
+
+    mdata->timer = NULL;
+    r = command_queue_process(data);
+    if (r < 0)
+        SOL_ERR("Error processing LCD's I2C command queue: %s\n",
+            sol_util_strerrora(r));
+
+    return false;
+}
 
 static int
 timer_reschedule(struct lcd_data *mdata,
     unsigned int timeout_ms,
-    bool (*cb)(void *data))
+    bool delete_prev)
 {
-    mdata->timer = sol_timeout_add(timeout_ms, cb, mdata);
+    if (mdata->timer) {
+        if (!delete_prev)
+            return 0;
+        else
+            sol_timeout_del(mdata->timer);
+    }
+
+    mdata->timer = sol_timeout_add(timeout_ms, timer_cb, mdata);
     SOL_NULL_CHECK(mdata->timer, -ENOMEM);
 
     return 0;
 }
 
 static void
-i2c_write_cb(void *cb_data, struct sol_i2c *i2c, uint8_t reg, uint8_t *data, ssize_t status)
+i2c_write_cb(void *cb_data,
+    struct sol_i2c *i2c,
+    uint8_t reg,
+    uint8_t *data,
+    ssize_t status)
 {
-    struct lcd_data *mdata = cb_data;
+    struct command *cmd = cb_data;
+    struct lcd_data *mdata = cmd->mdata;
 
     mdata->i2c_pending = NULL;
-    if (status < 0) {
-        SOL_ERR("Failed to process i2c command, no new commands will be executed.");
-        return;
-    }
+    SOL_INT_CHECK_GOTO(status, < 0, end);
+
+    if (cmd->value != LCD_CLEAR)
+        command_queue_process(mdata);
 
-    command_queue_process(cb_data);
+end:
+    free(cmd);
 }
 
-static bool
+static inline void
+command_copy(const struct command *src, struct command *dst)
+{
+    memcpy(dst, src, sizeof(struct command));
+}
+
+static int
 command_send(struct lcd_data *mdata, struct command *cmd)
 {
+    struct command *cpy;
+
     if (!sol_i2c_set_slave_address(mdata->i2c, cmd->chip_addr)) {
         SOL_WRN("Failed to set slave at address 0x%02x\n", cmd->chip_addr);
-        return false;
+        return -EIO;
     }
 
-    SOL_DBG("ChipAddr: 0x%02x, DataAddr: 0x%02x, Value 0x%02x - %c",
-        cmd->chip_addr, cmd->data_addr, cmd->value, cmd->value);
+    /* Unfortunately we can't afford not replicating this memory. At
+     * least &cmd->value must be alive during the whole callback time,
+     * and we can't guarantee that on our side with so many list
+     * operations going on. */
+    cpy = calloc(1, sizeof(*cpy));
```

Comment:
Use `malloc()` here, since the whole block is going to be overwritten anyway, there's no need to zero it out.